src/java/main/org/apache/zookeeper/server/ServerCnxn.java

```diff
@@ -51,7 +47,7 @@
     final public static Object me = new Object();
     private static final Logger LOG = LoggerFactory.getLogger(ServerCnxn.class);
     
-    protected ArrayList<Id> authInfo = new ArrayList<Id>();
+    private Set<Id> authInfo = Collections.newSetFromMap(new ConcurrentHashMap<Id, Boolean>());
```

Comment:
@fpj this is a more aggressive change but one for the better. 1) there's no good reason to have this field be protected; 2) it was defined as a List but used as a Set - so correct that; 3) make it concurrent/thread safe